buildscript/util.go

```diff
@@ -0,0 +1,205 @@
+	"bytes"
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/craiggwilson/goke/pkg/sh"
+	"github.com/craiggwilson/goke/task"
+	"github.com/pkg/errors"
+)
+
+func devBinDir() (string, error) {
+	return getRepoRootSubdir("dev-bin")
+}
+
+func devBinFile(filename string) (string, error) {
+	return getDirExeName("dev-bin", filename)
+}
+
+// getDirExeName returns the full path to the file in a subdirectory of the repo
+// root. For example: getDirExeName("dev-bin", "precious") returns
+// /path/to/mongosync/dev-bin/precious on unix-like systems and
+// C:\path\to\mongosync\dev-bin\precious.exe on Windows.
+func getDirExeName(dirname, filename string) (string, error) {
+	dir, err := getRepoRootSubdir(dirname)
+	if err != nil {
+		return "", err
+	}
+
+	target := filepath.Join(dir, filename)
+	if runtime.GOOS == "windows" {
+		target += ".exe"
+	}
+
+	return target, nil
+}
+
+// getRepoRootSubdir returns the absolute path to dirname inside the root of
+// the repository, creating it if it doesn't exist.
+func getRepoRootSubdir(dirname string) (string, error) {
+	root, err := repoRoot()
+	if err != nil {
+		return "", err
+	}
+
+	fullPath := filepath.Join(root, dirname)
+	err = os.MkdirAll(fullPath, 0755)
+
+	return fullPath, err
+}
+
+var repoRootDir string
+
+// repoRoot is a wrapper around git.FindRepoRoot(), but caches it so that
+// multiple calls in the same invocation of the program won't run git multiple
+// times.
+func repoRoot() (string, error) {
+	if repoRootDir != "" {
+		return repoRootDir, nil
+	}
+
+	var err error
+	repoRootDir, err = findRepoRoot(context.Background())
+	return repoRootDir, err
+}
+
+// findRepoRoot returns the filesystem path of the root of the git repository
+// that contains the process’s current directory.
+func findRepoRoot(ctx context.Context) (string, error) {
+	cmd := exec.CommandContext(ctx, "git", "rev-parse", "--show-toplevel")
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to find git repo root folder: %s", string(output))
+	}
+
+	return string(bytes.TrimSpace(output)), nil
+}
+
+func executableExistsWithVersion(ctx *task.Context, exe, exeVersion string) (bool, error) {
+	exists, err := fileExists(exe)
+	if err != nil {
+		return false, err
+	}
+	if exists {
+		var matches bool
+		matches, err = versionMatches(ctx, exe, exeVersion)
+		if err != nil {
+			return false, err
+		}
+		if matches {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func fileExists(p string) (bool, error) {
+	info, err := os.Stat(p)
+	if err == nil && info.Mode().IsRegular() {
+		return true, nil
+	}
+	if errors.Is(err, os.ErrNotExist) {
+		return false, nil
+	}
+	return false, err
+}
+
+func versionMatches(ctx *task.Context, exe, wantVersion string) (bool, error) {
+	out, err := sh.RunOutput(ctx, exe, "--version")
+	if err != nil {
+		return false, err
+	}
+	return strings.Contains(out, wantVersion), nil
```

Comment:
It's probably not going to be a problem anytime soon, but `wantVersion="0.7.2"` would match versions like `"0.7.21"`, `"10.7.2"`, etc.

Reply:
Good point. I changed this to use a regex with `\b` on either side of the version string.